File: android/app/src/main/java/com/aitutor/app/data/repository/AuthRepositoryImpl.kt
package com.aitutor.app.data.repository

import com.aitutor.app.data.remote.SupabaseProvider
import com.aitutor.app.domain.model.AppUser
import com.aitutor.app.domain.model.AuthState
import com.aitutor.app.domain.repository.AuthRepository
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.Google
import io.github.jan.supabase.auth.providers.builtin.Email
import io.github.jan.supabase.auth.status.SessionStatus
import io.github.jan.supabase.auth.user.UserInfo
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepositoryImpl @Inject constructor(
    supabase: SupabaseProvider
) : AuthRepository {

    private val auth = supabase.client.auth

    override val authState: Flow<AuthState> = auth.sessionStatus.map { status ->
        when (status) {
            is SessionStatus.Authenticated -> AuthState.SignedIn(status.session.user.toAppUser())
            is SessionStatus.NotAuthenticated -> AuthState.SignedOut
            else -> AuthState.Unknown   // LoadingFromStorage / NetworkError
        }
    }

    override suspend fun signInWithEmail(email: String, password: String): Result<Unit> =
        runCatching {
            auth.signInWith(Email) {
                this.email = email
                this.password = password
            }
        }

    override suspend fun signUpWithEmail(
        email: String,
        password: String,
        displayName: String?
    ): Result<Unit> = runCatching {
        auth.signUpWith(Email) {
            this.email = email
            this.password = password
            if (!displayName.isNullOrBlank()) {
                data = JsonObject(mapOf("display_name" to JsonPrimitive(displayName)))
            }
        }
    }

    override suspend fun signInWithGoogle(): Result<Unit> = runCatching {
        auth.signInWith(Google)
    }

    override suspend fun signOut(): Result<Unit> = runCatching { auth.signOut() }

    override suspend fun sendPasswordReset(email: String): Result<Unit> = runCatching {
        auth.resetPasswordForEmail(email)
    }

    private fun UserInfo?.toAppUser(): AppUser {
        require(this != null) { "UserInfo is null" }
        val meta = userMetadata
        val name = meta?.get("display_name")?.jsonPrimitive?.contentOrNull
        val avatar = meta?.get("avatar_url")?.jsonPrimitive?.contentOrNull
        return AppUser(
            id = id,
            email = email,
            displayName = name,
            avatarUrl = avatar
        )
    }
}

File: ios/AITutor/Data/Repository/AuthRepositoryImpl.swift
import Foundation
import Auth
import Supabase

final class AuthRepositoryImpl: AuthRepository {
    private let provider: SupabaseProvider

    init(provider: SupabaseProvider = .shared) {
        self.provider = provider
    }

    private var auth: AuthClient { provider.client.auth }

    var authState: AsyncStream<AuthState> {
        AsyncStream { continuation in
            let task = Task { [auth] in
                continuation.yield(.unknown)
                for await change in auth.authStateChanges {
                    let mapped: AuthState
                    switch change.event {
                    case .signedIn, .tokenRefreshed, .userUpdated, .initialSession:
                        if let session = change.session {
                            mapped = .signedIn(Self.mapUser(session.user))
                        } else {
                            mapped = .signedOut
                        }
                    case .signedOut, .userDeleted:
                        mapped = .signedOut
                    case .passwordRecovery, .mfaChallengeVerified:
                        // Recovery flow keeps the user on whatever screen they were on.
                        if let session = change.session {
                            mapped = .signedIn(Self.mapUser(session.user))
                        } else {
                            mapped = .signedOut
                        }
                    @unknown default:
                        mapped = .signedOut
                    }
                    continuation.yield(mapped)
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    func signInWithEmail(email: String, password: String) async throws {
        _ = try await auth.signIn(email: email, password: password)
    }

    func signUpWithEmail(email: String, password: String, displayName: String?) async throws {
        if let displayName, !displayName.isEmpty {
            _ = try await auth.signUp(
                email: email,
                password: password,
                data: ["display_name": .string(displayName)]
            )
        } else {
            _ = try await auth.signUp(email: email, password: password)
        }
    }

    func signInWithGoogle() async throws {
        try await auth.signInWithOAuth(
            provider: .google,
            redirectTo: URL(string: "aitutor://auth-callback")
        )
    }

    func signOut() async throws {
        try await auth.signOut()
    }

    func sendPasswordReset(email: String) async throws {
        try await auth.resetPasswordForEmail(email)
    }

    private static func mapUser(_ user: Auth.User) -> AppUser {
        let meta = user.userMetadata
        let displayName = meta["display_name"]?.stringValue
        let avatar = meta["avatar_url"]?.stringValue
        return AppUser(
            id: user.id.uuidString,
            email: user.email,
            displayName: displayName,
            avatarUrl: avatar
        )
    }
}

private extension AnyJSON {
    var stringValue: String? {
        if case let .string(s) = self { return s }
        return nil
    }
}
